### PR TITLE
CSI Plugin cleanup [1/3]: reduce global state

### PR DESCRIPTION
**Pull request by hoegaarden**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This is one of three PRs with the goal to remove some global state in the `csi` package.

This PR introduces the following important changes:
- global state in the `csi` package is removed (specifically: `csi.nim` & `csi.csiDrivers`); this state is moved onto the plugin struct
- the map of registered csi drivers has been moved into its own struct which internally manages the mutex
- the plugin implements the methods for both the `VolumePlugin` and the `PluginHandler` interfaces

Right now the kubelet still expects a separate VolumePlugin and a Registration Handler. To allow for that, we make sure to only ever create one instance of the `csi.plugin` and have make sure the VolumePlugin & the RegistrationHandler use this singleton. This is implemented by using a `sync.Once` and on package global variable holding the singleton instance.
**Note:** This singleton is only temporary and will go away with the last commit of this set, when the changes needed in the kubelet are figured out and merged.

**Which issue(s) this PR fixes**:

- Addresses parts of #69374
- Supersedes https://github.com/kubernetes/kubernetes/pull/69462

**Special notes for your reviewer**:

The whole set of PRs:
1. #70379
1. #70380
1. #70381

All the fakes we introduce for testing are made by the counterfeiter tool. A PR introducing update and verify scripts will be issued shortly.

**Does this PR introduce a user-facing change?**:

```release-note
CSI plugin cleanup: Remove global state
```

/cc @vladimirvivien 
/cc @mariantalla 
/cc @astrieanna
/cc @benmoss

### CODE REVIEW

File: pkg/volume/csi/csi_plugin.go [hunk 1]
```diff
@@ -60,6 +60,8 @@ const (
 	csiResyncPeriod = time.Minute
 )
 
+var errNotInitialized = errors.New("CSI plugin not initilaized")
```
Comment (miguelbernadi) [inline pkg/volume/csi/csi_plugin.go]: Typo: `initialized`

**Comment by vladimirvivien** [comment]
@hoegaarden Now that 1.13 is stable, will you resume work on this set of PRs ?

**Comment by hoegaarden** [comment]
/close
superseded by #73919

**Pull request closed without merging**